lib/compress/zstd_double_fast.c

```diff
@@ -312,8 +359,12 @@ size_t ZSTD_compressBlock_doubleFast_dictMatchState_generic(
         U32 offset;
         size_t const h2 = ZSTD_hashPtr(ip, hBitsL, 8);
         size_t const h = ZSTD_hashPtr(ip, hBitsS, mls);
-        size_t const dictHL = ZSTD_hashPtr(ip, dictHBitsL, 8);
-        size_t const dictHS = ZSTD_hashPtr(ip, dictHBitsS, mls);
+        size_t const dictHashAndTagL = ZSTD_hashPtr(ip, dictHBitsL + ZSTD_SHORT_CACHE_TAG_BITS, 8);
+        size_t const dictHashAndTagS = ZSTD_hashPtr(ip, dictHBitsS + ZSTD_SHORT_CACHE_TAG_BITS, mls);
```

Comment:
These should be `U32`s, right? Well ok, I guess the max hashLog is like 30 and if you add 8 it spills over... but if we're limiting the dict content to 16MB there's no point in having gigantic tables. So maybe it's worth capping the hashLog for CDicts for these levels?

Even if you don't do that, using a `size_t` is going to do funky things on a 32-bit platform...

Reply:
`ZSTD_hashPtr` calls in `zstd_fast.c` and `zstd_double_fast.c` currently all use `size_t`. If we want to change that, it should be a separate PR and applied uniformly across all matchfinders.
```
-        size_t const dictHL = ZSTD_hashPtr(ip, dictHBitsL, 8);
-        size_t const dictHS = ZSTD_hashPtr(ip, dictHBitsS, mls);
+        size_t const dictHashAndTagL = ZSTD_hashPtr(ip, dictHBitsL + ZSTD_SHORT_CACHE_TAG_BITS, 8);
+        size_t const dictHashAndTagS = ZSTD_hashPtr(ip, dictHBitsS + ZSTD_SHORT_CACHE_TAG_BITS, mls);
```

Reply:
@felixhandte all of hash functions return `size_t`.

However, good point about the large hash log. We need to make sure that the dictionaries hash log is <= 24.

Elliot added asserts for that, but we need to make sure that the dictionaries hash log is <= 24. We need to do the same for the row based match finder. I was able to trigger the asserts with the row-based match finder.

Reply:
E.g. `./zstd -o /dev/null -7 < ~/datasets/enwik7 --zstd=wlog=29,hlog=29`